collector/perf_linux.go

```diff
@@ -0,0 +1,549 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+
+	perf "github.com/hodgesds/perf-utils"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+const (
+	perfSubsystem = "perf"
+)
+
+var (
+	perfHwProfilers    = map[int]perf.HardwareProfiler{}
+	perfSwProfilers    = map[int]perf.SoftwareProfiler{}
+	perfCacheProfilers = map[int]perf.CacheProfiler{}
+	perfInitializeOnce sync.Once
+)
+
+func init() {
+	registerCollector(perfSubsystem, defaultDisabled, NewPerfCollector)
+}
+
+// perfCollector is a Collected that uses the perf subsystem to collect
+// metrics. It uses perf_event_open an ioctls for profiling.
+type perfCollector struct{}
+
+// initializePerfProfilers is used to setup
+func initializePerfProfilers() {
+	ncpus := runtime.NumCPU()
+	for i := 0; i < ncpus; i++ {
+		// Use -1 to profile all processes on the CPU, see:
+		// man perf_event_open
+		perfHwProfilers[i] = perf.NewHardwareProfiler(-1, i)
+		if err := perfHwProfilers[i].Start(); err != nil {
+			log.Errorf(
+				"failed to start perf hardware profiler on CPU (%d): %+v",
+				i,
+				err)
+		}
+		perfSwProfilers[i] = perf.NewSoftwareProfiler(-1, i)
+		if err := perfSwProfilers[i].Start(); err != nil {
+			log.Errorf(
+				"failed to start perf software profiler on CPU (%d): %+v",
+				i,
+				err)
+		}
+		perfCacheProfilers[i] = perf.NewCacheProfiler(-1, i)
+		if err := perfCacheProfilers[i].Start(); err != nil {
+			log.Errorf(
+				"failed to start perf cache profiler on CPU (%d): %+v",
+				i,
+				err)
+		}
+	}
+}
+
+// NewPerfCollector returns a new perf based collector, it creates a profiler
+// per CPU.
+func NewPerfCollector() (Collector, error) {
+	perfInitializeOnce.Do(initializePerfProfilers)
+	return &perfCollector{}, nil
+}
+
+func (c *perfCollector) Update(ch chan<- prometheus.Metric) error {
+	if err := c.updateHardwareStats(ch); err != nil {
+		return err
+	}
+
+	if err := c.updateSoftwareStats(ch); err != nil {
+		return err
+	}
+
+	if err := c.updateCacheStats(ch); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *perfCollector) updateHardwareStats(ch chan<- prometheus.Metric) error {
+	for cpu, profiler := range perfHwProfilers {
+		cpuStr := fmt.Sprintf("%d", cpu)
+		hwProfile, err := profiler.Profile()
+		if err != nil {
+			return err
+		}
+		if hwProfile == nil {
+			continue
+		}
+
+		if hwProfile.CPUCycles != nil {
```

Comment:
Would be nice to refactor this a bit, maybe by using a map and loop over it like we do in similar cases in other collectors.

Reply:
I made a couple of attempts at this but creating the map becomes rather difficult without using reflection because not all struct fields may be present. I can git it another attempt and it ***might*** save a little of the redundant checks, but would probably be slower. What do you think?

Reply:
I think I got it working pretty well now.